File: core/models/contact_sync.py
from sqlalchemy import Column, String, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID, JSONB

from core.models.base import OrgScopedModel


class ContactSync(OrgScopedModel):
    """One run of the datasource-driven import pipeline into a directory.

    Records the mapping schema, optional CSV upload, optional auto-assign ``agent_id``,
    lifecycle ``status`` (pending|processing|completed|failed), per-run ``counts``
    ({created,updated,skipped,failed}) and per-row ``row_errors`` for a downloadable
    partial-failure report. ``directory_id``/``datasource_id`` are SET NULL so a sync
    survives as detached audit when its directory is hard-deleted.
    """

    __tablename__ = "contact_syncs"

    directory_id = Column(
        UUID(as_uuid=True),
        ForeignKey("contact_directories.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    datasource_id = Column(
        UUID(as_uuid=True),
        ForeignKey("datasources.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    schema_id = Column(
        UUID(as_uuid=True),
        ForeignKey("contact_schemas.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    upload_id = Column(
        UUID(as_uuid=True),
        ForeignKey("uploads.id", ondelete="SET NULL"),
        nullable=True,
    )
    agent_id = Column(
        UUID(as_uuid=True),
        ForeignKey("agents.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    # pending | processing | completed | failed
    status = Column(String(20), nullable=False, default="pending", index=True)
    counts = Column(JSONB, nullable=False, default=dict)  # {created,updated,skipped,failed}
    row_errors = Column(JSONB, nullable=False, default=list)  # [{row,field,message}]
    error = Column(String(500), nullable=True)  # hard-failure message
    # Optional external integration references (e.g. a future REST/app datasource sync).
    app_id = Column(String(255), nullable=True)
    connection_id = Column(String(255), nullable=True)
    created_by_user_id = Column(UUID(as_uuid=True), nullable=True)

    def to_dict(self) -> dict:
        return {
            "id": str(self.id),
            "directory_id": str(self.directory_id) if self.directory_id else None,
            "datasource_id": str(self.datasource_id) if self.datasource_id else None,
            "schema_id": str(self.schema_id) if self.schema_id else None,
            "upload_id": str(self.upload_id) if self.upload_id else None,
            "agent_id": str(self.agent_id) if self.agent_id else None,
            "app_id": self.app_id,
            "connection_id": self.connection_id,
            "status": self.status,
            "counts": self.counts or {},
            "row_errors": self.row_errors or [],
            "error": self.error,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: frontend/src/types/contactSync.ts
/**
 * ContactSync types — mirrors `core/models/contact_sync.py` `to_dict()` and the
 * contact-syncs router shapes. A `ContactSync` is one run of the datasource-driven CSV
 * import into a directory (partial-success: valid rows imported, invalid rows recorded).
 */

/** Sync lifecycle status (pending → processing → completed | failed). */
export type ContactSyncStatus = 'pending' | 'processing' | 'completed' | 'failed';

/** The terminal statuses — polling stops once a sync reaches one of these. */
export const CONTACT_SYNC_TERMINAL_STATUSES: readonly ContactSyncStatus[] = ['completed', 'failed'];

/** Per-run counts written by `run_contact_sync`. */
export interface ContactSyncCounts {
  created?: number;
  updated?: number;
  skipped?: number;
  failed?: number;
}

/** A single per-row error entry in `row_errors` (partial-failure detail). */
export interface ContactSyncRowError {
  row: number;
  field: string | null;
  message: string;
  /** The failed row's identity, so the report shows WHICH contact failed. */
  name?: string | null;
  phone_number?: string | null;
}

/** A sync as returned by `to_dict()` (create / get / list rows). */
export interface ContactSync {
  id: string;
  directory_id: string | null;
  datasource_id: string | null;
  schema_id: string | null;
  upload_id: string | null;
  agent_id: string | null;
  status: ContactSyncStatus;
  counts: ContactSyncCounts;
  row_errors: ContactSyncRowError[];
  error: string | null;
  created_at: string | null;
  updated_at: string | null;
}

/**
 * Multipart body for `POST /contact-syncs`. Sent as `FormData` (the CSV `file` plus the
 * scalar fields), NOT JSON.
 */
export interface CreateContactSyncPayload {
  directory_id: string;
  schema_id: string;
  agent_id?: string | null;
  file: File;
}

/** Extra scoping field for `POST /contact-syncs/list` (on top of paging params). */
export interface ListContactSyncsParams {
  directory_id?: string | null;
  status?: ContactSyncStatus | null;
}
